### PR TITLE
Reduce enthusiasm for calling flush APIs

### PR DESCRIPTION
**Pull request by DaveCTurner**
Today we only really describe the benefits of the flush and synced-flush APIs:

> _... reduces recovery times ..._

> _... NOTE: It is harmless to request a synced flush ..._

This might lead users to believe that they should call these APIs frequently as
a matter of course, but this is not what we recommend.

This change rewords the docs a bit to talk about trade-offs and to mention more
prominently that Elasticsearch does these things automatically.

### CODE REVIEW

File: docs/reference/indices/flush.asciidoc [hunk 1]
```diff
@@ -1,13 +1,16 @@
...
-flush process of an index makes sure that any data that is currently only
-persisted in the <<index-modules-translog,transaction log>> is also permanently
-persisted in Lucene. This reduces recovery times as that data doesn't need to be
-reindexed from the transaction logs after the Lucene indexed is opened. By
-default, Elasticsearch uses heuristics in order to automatically
-trigger flushes as required. It is rare for users to need to call the API directly.
+The flush process of an index makes sure that any data that is currently only
+stored in the <<index-modules-translog,transaction log>> is also permanently
+stored in Lucene. When restarting, {es} replays any unflushed operations from
+the transaction log into Lucene to ensure that searches do not return stale
```
Comment (bleskes) [inline docs/reference/indices/flush.asciidoc]: I liked the previous explanation better 👼 The real reason is that we want to bring lucene back to where it was before the shard was shutdown and lucene resets itself to the last time it was flushed. Stale searches are only one aspect of it.

File: docs/reference/indices/flush.asciidoc [hunk 2]
```diff
@@ -24,21 +27,21 @@ The flush API accepts the following request parameters:
 
 [horizontal]
 `wait_if_ongoing`::  If set to `true` the flush operation will block until the
-flush can be executed if another flush operation is already executing.
-The default is `false` and will cause an exception to be thrown on
-the shard level if another flush operation is already running.
+flush can be executed if another flush operation is already executing. The
```
Comment (bleskes) [inline docs/reference/indices/flush.asciidoc]: the main point is here that the command waits until a flush has happened and all operations indexed before the command was called are guaranteed to be flushed. I'm tempting to also say we need to change the default, but that's a different discussion :)

File: docs/reference/indices/flush.asciidoc [hunk 3]
```diff
@@ -122,26 +126,25 @@ which returns something similar to:
...
-the synced flush to fail on that shard. This means that some shards may be synced flushed while others aren't. See below for more.
-2. The `sync_id` marker is removed as soon as the shard is flushed again. That is because a flush replaces the low level
-lucene commit point where the marker is stored. Uncommitted operations in the transaction log do not remove the marker.
-In practice, one should consider any indexing operation on an index as removing the marker as a flush can be triggered by Elasticsearch
-at any time.
-
-
-NOTE: It is harmless to request a synced flush while there is ongoing indexing. Shards that are idle will succeed and shards
- that are not will fail. Any shards that succeeded will have faster recovery times.
+The Synced Flush API allows an administrator to initiate a synced flush
```
Comment (bleskes) [inline docs/reference/indices/flush.asciidoc]: Why did you change the old explanation?

File: docs/reference/indices/flush.asciidoc [hunk 4]
```diff
@@ -52,26 +55,27 @@ POST _flush
 [[indices-synced-flush]]
 === Synced Flush
 
-Elasticsearch tracks the indexing activity of each shard. Shards that have not
-received any indexing operations for 5 minutes are automatically marked as inactive. This presents
-an opportunity for Elasticsearch to reduce shard resources and also perform
-a special kind of flush, called `synced flush`. A synced flush performs a normal flush, then adds
-a generated unique marker (sync_id) to all shards.
+{es} keeps track of which shards have received indexing activity recently, and
```
Comment (bleskes) [inline docs/reference/indices/flush.asciidoc]: as in the previous comment - I like the previous explanation better. It explains what the thing does. With the new version, it provides a higher level description that to me is harder to understand what really happens.

**Review by bleskes — COMMENTED** [review]
I left some comments. I general I think it's no surprise that the old docs read better to me, because I wrote them :) That said - I think this PRs tries to do too much in one go. If you want to re-write the whole page, that's fine with me but it seems the goal was to only make it clearer that (normal) flushes should be very rarely called.

**Comment by javanna** [comment]
what should we do with this PR? It was opened months ago and there was some discussion, hence it was not merged. Should we close it or get it in?

**Comment by DaveCTurner** [comment]
I am holding off on further changes here right now while the work on history retention (#41536) is still in flight, because that is intertwined with the flushing and recovery mechanisms. This PR is still on my list.

**Comment by DaveCTurner** [comment]
Closing this in favour of #46245. @bleskes I will reply to your comments here, having addressed some of them in #46245.

File: docs/reference/indices/flush.asciidoc (hunk 1 above)
Comment (DaveCTurner) [reply docs/reference/indices/flush.asciidoc]: That's ok with me - see updated wording in https://github.com/elastic/elasticsearch/pull/46245/files#diff-31f3afe6b50f556a95a396324be535e5R6

File: docs/reference/indices/flush.asciidoc (hunk 2 above)
Comment (DaveCTurner) [reply docs/reference/indices/flush.asciidoc]: Ok. This wasn't spelled out before as far as I can see, but it's a good point to make. See https://github.com/elastic/elasticsearch/pull/46245/files#diff-31f3afe6b50f556a95a396324be535e5R22-R24.

The default is `true`, fixed by Nhat recently.

File: docs/reference/indices/flush.asciidoc (hunk 4 above)
Comment (DaveCTurner) [reply docs/reference/indices/flush.asciidoc]: I don't follow you here. I've perhaps reordered the information to start with the higher-level view and then drop down to the details, since this is easier to follow if you aren't already intimately familiar with it. I don't think I've lost any detail here have I?

File: docs/reference/indices/flush.asciidoc (hunk 3 above)
Comment (DaveCTurner) [reply docs/reference/indices/flush.asciidoc]: The changes I made were to address a number of points that I originally found confusing:

- It says it's useful specifically for rolling restarts, but it's actually useful for all restarts whether rolling or otherwise.
- The note about future flushes removing the `sync_id` is out of place in this section on the API, since it applies even if you don't use this API.
- It introduces the concept of a "low level Lucene commit point" without definition or explanation. This isn't really necessary, we can make that same point more clearly by talking about flushes directly.
- The formatting (i.e. the numbered list and the `NOTE:` callout) were unnecessary and made it harder to read than just straight prose.
